packages/financial-templates-lib/src/clients/InsuredBridgeL1Client.ts

```diff
@@ -0,0 +1,181 @@
+  recipient: string;
+  l1Token: string;
+  amount: string;
+  slowRelayFeePct: string;
+  instantRelayFeePct: string;
+  realizedLpFeePct: string;
+  priceRequestAncillaryDataHash: string;
+  depositHash: string;
+  depositContract: string;
+  relayState: relayState;
+}
+
+export class InsuredBridgeL1Client {
+  public readonly bridgeAdmin: BridgeAdminWeb3;
+  public bridgePools: { [key: string]: BridgePoolWeb3 }; // L1TokenAddress=>BridgePoolClient
+
+  private relays: { [key: string]: { [key: string]: Relay } } = {}; // L1TokenAddress=>depositHash=>Relay.
+
+  private firstBlockToSearch: number;
+
+  constructor(
+    private readonly logger: Logger,
+    readonly bridgeAdminAbi: Abi,
+    readonly bridgePoolAbi: Abi,
+    readonly l1Web3: Web3,
+    readonly bridgeAdminAddress: string,
+    readonly startingBlockNumber: number = 0,
+    readonly endingBlockNumber: number | null = null
+  ) {
+    this.bridgeAdmin = (new l1Web3.eth.Contract(
+      bridgeAdminAbi,
+      bridgeAdminAddress
+    ) as unknown) as InsuredBridgeL1Client["bridgeAdmin"]; // Cast to web3-specific type
+
+    this.bridgePools = {}; // Initialize the bridgePools with no pools yet. Will be populated in the _initialSetup.
+
+    this.firstBlockToSearch = startingBlockNumber;
+  }
+
+  // Return an array of all bridgePool addresses
+  getBridgePoolsAddresses() {
+    this._throwIfNotInitialized();
+    return Object.values(this.bridgePools).map((bridgePool: BridgePoolWeb3) => bridgePool.options.address);
+  }
+
+  getAllRelayedDeposits() {
+    this._throwIfNotInitialized();
+    return Object.keys(this.relays)
+      .map((l1Token: string) =>
+        Object.keys(this.relays[l1Token]).map((depositHash: string) => this.relays[l1Token][depositHash])
+      )
+      .flat();
+  }
+
+  getRelayedDepositsForL1Token(l1Token: string) {
+    this._throwIfNotInitialized();
+    return Object.values(this.relays[l1Token]);
+  }
+
+  async _initialSetup(fromBlock: number): Promise<void> {
+    const whitelistedTokenEvents = await this.bridgeAdmin.getPastEvents("WhitelistToken", { fromBlock });
+    for (const whitelistedTokenEvent of whitelistedTokenEvents) {
+      const l1Token = whitelistedTokenEvent.returnValues.l1Token;
+      this.bridgePools[l1Token] = (new this.l1Web3.eth.Contract(
+        this.bridgePoolAbi,
+        whitelistedTokenEvent.returnValues.bridgePool
+      ) as unknown) as BridgePoolWeb3;
+      this.relays[l1Token] = {};
+    }
+  }
+
+  async update(): Promise<void> {
+    // Define a config to bound the queries by.
+    const blockSearchConfig = {
+      fromBlock: this.firstBlockToSearch,
+      toBlock: this.endingBlockNumber || (await this.l1Web3.eth.getBlockNumber()),
+    };
+
+    if (!this._isClientInitialized()) await this._initialSetup(blockSearchConfig.fromBlock);
+
+    // Fetch event information
+    // TODO: consider optimizing this further. Right now it will make a series of sequential BlueBird calls for each pool.
```

Comment:
Aren't these calls made in parallel?

Reply:
the fetches within the `Promise.all` are parallel but there are a sequential set of parallel calls for each `bridgePool`. Put differently, if there are 5 bridge pools then there are 5 sequential sets of 4 parallel calls.

Reply:
agreed

Reply:
comment was a bit confusing.